File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.py
from heapq import heapify, heappop, heappush
def solution(operations):
    max_heap = []
    min_heap = []
    
    for operation in operations:
        operation = operation.split(" ")
        o = operation[0]
        v = int(operation[1])
        
        if o == 'I':    # 삽입
            heappush(max_heap,-v)
            heappush(min_heap,v)
        elif o == "D" and len(min_heap)!=0:   # 삭제
            if v == 1:
                max_value = heappop(max_heap)
                min_heap.remove(-max_value)
            else:
                min_value = heappop(min_heap)
                max_heap.remove(-min_value)

    return  [0,0] if len(min_heap)==0 else [-heappop(max_heap),heappop(min_heap)]

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.js
function solution(operations) {
    const heap = [];
    const op = operations.map(operation => operation.split(' '));
  	// 입력된 명령어를 공백(' ')을 기준으로 분할한다.
  	// 따라서 배열[0] = 명령어, 배열[1] = 숫자로 접근할 수 있다.
    
    op.forEach(num => {
        if(num[0] === 'I') {	// 명령어가 I라면 데이터 삽입
            heap.push(Number(num[1]))
        }
        else {	// 그 외의 경우, 즉 명령어가 D인 경우
            const findValue = (num[1] === '1' ? Math.max : Math.min)(...heap);
            const delIdx = heap.indexOf(findValue);
            heap.splice(delIdx, 1);
        }
    })
    
    return heap.length ? [Math.max(...heap), Math.min(...heap)] : [0, 0];
}